File: library/src/main/java/com/gituhb/dmitrikudrenko/countryphonecodechooser/model/CountryCode.java
package com.gituhb.dmitrikudrenko.countryphonecodechooser.model;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import com.google.gson.annotations.SerializedName;

public class CountryCode implements Comparable<CountryCode>, Parcelable {
    @SerializedName("name")
    private String name;
    @SerializedName("dial_code")
    private String phoneCode;
    @SerializedName("code")
    private String code;
    @SerializedName("is_default")
    private boolean isDefault;

    private CountryCode() {
    }

    public CountryCode(String name, String phoneCode, String code) {
        this.name = name;
        this.phoneCode = phoneCode;
        this.code = code;
    }

    @Override
    public int compareTo(@NonNull CountryCode countryCode) {
        return name.compareTo(countryCode.name);
    }

    public String getName() {
        return name;
    }

    public String getCode() {
        return code.toLowerCase();
    }

    public String getPhoneCode() {
        return phoneCode;
    }

    public String getPhoneCodeNonSpaced() {
        return phoneCode.replaceAll(" ", "");
    }

    public boolean isDefault() {
        return isDefault;
    }

    public char getSection() {
        return name.charAt(0);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeString(name);
        parcel.writeString(phoneCode);
        parcel.writeString(code);
        parcel.writeInt(isDefault ? 1 : 0);
    }

    public static final Creator<CountryCode> CREATOR = new Creator<CountryCode>() {
        @Override
        public CountryCode createFromParcel(Parcel source) {
            CountryCode countryCode = new CountryCode();
            countryCode.name = source.readString();
            countryCode.phoneCode = source.readString();
            countryCode.code = source.readString();
            countryCode.isDefault = source.readInt() == 1;
            return countryCode;
        }

        @Override
        public CountryCode[] newArray(int size) {
            return new CountryCode[size];
        }
    };

    @Override
    public boolean equals(Object o) {
        return (o instanceof CountryCode) && name.equals(((CountryCode) o).name)
                && phoneCode.equals(((CountryCode) o).phoneCode);
    }
}

File: library/src/main/kotlin/com/github/dmitrikudrenko/countryphonecodechooser/model/CountryCode.kt
package com.github.dmitrikudrenko.countryphonecodechooser.model

import android.os.Parcel
import com.github.dmitrikudrenko.countryphonecodechooser.utils.KParcelable
import com.github.dmitrikudrenko.countryphonecodechooser.utils.parcelableCreator
import com.google.gson.annotations.SerializedName

class CountryCode private constructor(
        @SerializedName("name")
        var name: String = "",
        @SerializedName("dial_code")
        var phoneCode: String = "",
        @SerializedName("code")
        var code: String = "",
        @SerializedName("is_default")
        private var isDefault: Boolean = false
) : Comparable<CountryCode>, KParcelable {

    val section: Char?
        get() = name[0]

    override fun compareTo(other: CountryCode): Int {
        return name.compareTo(other.name)
    }

    override fun writeToParcel(dest: Parcel, flags: Int) = with(dest) {
        writeString(name)
        writeString(phoneCode)
        writeString(code)
        writeInt(if (isDefault) 1 else 0)
    }

    override fun equals(other: Any?): Boolean {
        return (other is CountryCode && name == other.name
                && phoneCode == other.phoneCode)
    }

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + phoneCode.hashCode()
        result = 31 * result + code.hashCode()
        result = 31 * result + isDefault.hashCode()
        return result
    }

    private constructor(p: Parcel) : this(
            p.readString(),
            p.readString(),
            p.readString(),
            p.readInt() == 1
    )

    companion object {
        @JvmField
        val CREATOR = parcelableCreator(::CountryCode)
    }
}
